### ISSUE TITLE
.babelrc

### ISSUE THREAD

**Issue by sprlwrksAprilmintacpineda** [infernojs/create-inferno-app#76]
## Abstract

Babel helpers, like `_extends`, causes duplications when you use them everywhere on different files then import those files on another file that uses it, this actually causes duplicate codes on the build, I checked it and saw these on `main.js`

```jsx
var s = Object.assign || function (e) {
      for (var t = 1; t < arguments.length; t++) {
        var n = arguments[t];
        for (var r in n) Object.prototype.hasOwnProperty.call(n, r) && (e[r] = n[r])
      }
      return e
    },
```

```jsx
var a = Object.assign || function (e) {
      for (var t = 1; t < arguments.length; t++) {
        var n = arguments[t];
        for (var r in n) Object.prototype.hasOwnProperty.call(n, r) && (e[r] = n[r])
      }
      return e
    },
```

I simply did a search for **Object.assign ||** and saw that there where 6 occurences on `main.js` alone! This is a big no-no, because in production these redundancies become dead codes that consumes resources but are not in used (actually they are in used but different places that use them could have referred to `s` itself and not create a new binding for it).

I researched and found out that babel devs themselves are aware of this and I also found this module https://babeljs.io/docs/en/babel-plugin-transform-runtime.

> Babel uses very small helpers for common functions such as _extend. By default this will be added to every file that requires it. This duplication is sometimes unnecessary, especially when your application is spread out over multiple files. This is where the transform-runtime plugin comes in: all of the helpers will reference the module babel-runtime to avoid duplication across your compiled output. The runtime will be compiled into your build. ~ https://babeljs.io/docs/en/babel-plugin-transform-runtime

> runtime will be compiled into your build

Sounds a bit scary since I don't know how big the module is, I'm actually looking for something better, maybe a smarter transpiler that knows not to redundantly repeat stuff and would help in eliminating dead codes to save our users from us (LOL 🤣 ).

## Objective

I am aiming to save some file size and eliminate dead codes on build as much as possible.

**Comment by Havunen** [comment]
If you want optimal output, you might want to consider rollup IIFE format + custom set of plugins (no preset) for babel. This way you get only once per application those helpers inlined. Also "typeof" babel plugin creates a lot of overhead. We simply removed it, but then you cannot use typeof for Symbols. I'm not sure if we can apply those patterns into this create-inferno-app, but you can definitely roll your own.

**Comment by sprlwrksAprilmintacpineda** [comment]
hmmm, if I use rollup and custom babel config, that would mean I would need to eject, right?

**Comment by Havunen** [comment]
@sprlwrksAprilmintacpineda yeah, it might be easier to just do everything yourself.. :/

**Comment by Havunen** [comment]
I will make a inferno starter project based on rollup and stuff, we use in our company... but there is so much to do atm. So that has lower priority

**Comment by aprilmintacpineda** [comment]
@Havuven maybe we could just do that with inferno-scripts?

**Comment by Havunen** [comment]
I have rebased all updates from the upstream, updated all the depencies to latest versions and converted the source code to ESM.
The next major version will require Node 20 or newer.

This issue is most likely fixed after all those changes to babel plugins. Closing.
